### DISCUSSION TITLE
Dotted lines, phantom lines are not printed from PDF files (if used with --no-sandbox)

### DISCUSSION THREAD

**Discussion by emitclever** [cefsharp/CefSharp#3776] [General]
## Bug Report
It is my first bug on github, I will try to explain simple as possible.

- **What version of the product are you using? *92.0.260*
Description.
Initially we had using with success the 90.6.7 release but lately printing not working for dotted lines. Our application had no suffered any changes. Using Edge, Adobe Reader the dotted lines are printed but no trough CefSharp browser.
Additional: If we use to print with Chrome Version 92.0.4515.159 (Official Build) (64-bit) then these lines print out correctly.

initial image (as seen trough cef webbrowser)
![image](https://user-images.githubusercontent.com/11360303/131148370-08edf63d-3ced-40f2-a471-be20e687395b.png)

after printing trough CEF webbrowser (scanned for results)
![image](https://user-images.githubusercontent.com/11360303/131148648-ee8c793a-5cd8-4414-8538-65ac77e58885.png)

- **What architecture x86 or x64?**
   x64
- **What version of .Net?**
4.5.2
    
- **On what operating system?**
Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WinForms

![image](https://user-images.githubusercontent.com/11360303/131149722-33825402-70f8-4dd6-9042-46aff83c70ad.png)

![image](https://user-images.githubusercontent.com/11360303/131161233-1b1293fe-0215-4eca-b2e9-1a4c1e226784.png)

also tested some parameters without success
![image](https://user-images.githubusercontent.com/11360303/131163183-992fb01b-7ae6-42d1-bc53-060ec79be727.png)

**Comment by amaitland** [comment]
I'm converting this to a discussion as PDF printing isn't implemented in CefSharp and is not something that we can fix directly. 

> Initially we had using with success the 90.6.7 release but lately printing not working for dotted lines. Our application had no suffered any changes. Using Edge, Adobe Reader the dotted lines are printed but no trough CefSharp browser.
Additional: If we use to print with Chrome Version 92.0.4515.159 (Official Build) (64-bit) then these lines print out correctly.

As per the bug report temple, please test with the latest stable version before opening an issue. Version 92.0.260 is based on Chromium 92.0.4515.159

**Reply by emitclever** [reply]
Thanks for input. I've already tested with the later version as related on additional section and does not work. Interesting is that with the latest version of Chrome browser on wich cef is build lately the pdf printing sems to be fixed. I have tried some arguments (cef.commandlines.args.add) but without success.

**Reply by amaitland** [reply]
Please test with the CEF Sample application. 

https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef

**Reply by emitclever** [reply]
using cef_binary_92.0.26+g9d87c1f+chromium-92.0.4515.159_windows64_client as application the dotted lines are printed ok. I cannot run the visual on example code too many errors. I still figuring out trough debugger.

**Reply by emitclever** [reply]
I update this and I apologize to use your time I had used the SharpBrowser project which is similar with this one.

**Reply by amaitland** [reply]
You can test with https://github.com/cefsharp/CefSharp.MinimalExample

Restore the Nuget packages, close visual studio, reopen and then rebuild.

**Comment by emitclever** [comment]
I just tried the minimal build as you explained to me and the results are similar dotted lines are out off scale and missplaced. I will put a pdf as example. I have thousands of pdf in my library.Thank you for help,
-------- Original message --------From: Alex Maitland ***@***.***> Date: 8/30/21  16:08  (GMT-05:00) To: cefsharp/CefSharp ***@***.***> Cc: emitclever ***@***.***>, Author ***@***.***> Subject: Re: [cefsharp/CefSharp] Dotted lines, phantom lines are not printed
  from PDF files (#3776) 
You can test with https://github.com/cefsharp/CefSharp.MinimalExample
Restore the Nuget packages, close visual studio, reopen and then rebuild.

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/cefsharp/CefSharp/discussions/3776#discussioncomment-1255975",
"url": "https://github.com/cefsharp/CefSharp/discussions/3776#discussioncomment-1255975",
"name": "View Discussion"
},
"description": "View this Discussion on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Reply by emitclever** [reply]
I had uses the minimal example compiled as explained. This is an example with dotted lines. Nothing special with this PDF all my technical files are similar in specs. With Chrome Version 92.0.4515.159 (Official Build) (64-bit) the lines are printed ok
[SHIPP-BRK-01___0.pdf](https://github.com/cefsharp/CefSharp/files/7084613/SHIPP-BRK-01___0.pdf)

**Reply by emitclever** [reply]
Tested also the WPF same thing, shown ok, the lines are printed off scale.
![image](https://user-images.githubusercontent.com/11360303/131577209-54192386-9476-4c6c-a714-ce690d5261d1.png)

**Reply by amaitland** [reply]
What command line args did you use to test cefclient?

**Reply by emitclever** [reply]
none

**Reply by emitclever** [reply]
also tested with minimal offscreen mode. Printed with same result. No arguments.
 const string testUrl = "M:\\Production_PDFs\\SHIPP-BRK-02___0.pdf";
....
 task.ContinueWith(x =>
                    {
                        browser.Print();

**Reply by amaitland** [reply]
Please test with the CEF Sample application as detailed in https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef

Make sure you use the command line arguments listed for WinForms.

**Comment by emitclever** [comment]
tested with x32 and x64 from  https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef with same printing results
![image](https://user-images.githubusercontent.com/11360303/131579562-cf474b7f-526f-4dfc-84ad-7c1a7b48aefd.png)

and results  are below
![image](https://user-images.githubusercontent.com/11360303/131580060-71ae2a90-3ce8-420b-bbdd-1380b5f71be3.png)

**Reply by amaitland** [reply]
> tested with x32 and x64 from https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef with same printing results

So the problem reproduces with cefclient?

**Comment by emitclever** [comment]
Quite correct. The chrome browser work well the only comparison that I could make at this moment. 
-------- Original message --------From: Alex Maitland ***@***.***> Date: 8/31/21  18:43  (GMT-05:00) To: cefsharp/CefSharp ***@***.***> Cc: "Adrian L." ***@***.***>, Author ***@***.***> Subject: Re: [cefsharp/CefSharp] Dotted lines, phantom lines are not printed
  from PDF files (#3776) 

tested with x32 and x64 from https://github.com/cefsharp/CefSharp/blob/master/CONTRIBUTING.md#cefsharp-vs-chromium-embedded-frameworkcef with same printing results

So the problem reproduces with cefclient?

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/cefsharp/CefSharp/discussions/3776#discussioncomment-1262494",
"url": "https://github.com/cefsharp/CefSharp/discussions/3776#discussioncomment-1262494",
"name": "View Discussion"
},
"description": "View this Discussion on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Reply by amaitland** [reply]
Thanks for clarifying. Can you also confirm that it works correctly when you run cefclient without command line arguments?

CEF (upstream) support for the PDF viewer currently differs from chrome. That is expected to change when https://bugs.chromium.org/p/chromium/issues/detail?id=702993 is implemented, this will hopefully be fixed then.

If the meantime if you can find a combination of command line args that cefclient works correctly with then it might be possible to work around the problem. I'd first try removing the no-sandbox arg, I vaguely remember another PDF related problem that only occurred when the sandbox was disabled.

**Reply by emitclever** [reply]
I just tested cefclient.exe without command line args and the document is printed correctly from my point of view.

initial
![image](https://user-images.githubusercontent.com/11360303/131673185-58978795-4a6c-4af9-ad68-82ca398e43db.png)
printed
![image](https://user-images.githubusercontent.com/11360303/131673245-09b04061-ff37-462f-b88b-70776edf1560.png)

**Reply by emitclever** [reply]
From now on I do not know from where to continue the integration or annotated it. The cefclient.exe without args seems working. The minimal code does not.

**Reply by amaitland** [reply]
You need to isolate which command line argument causes the problem with cefclient.

**Comment by emitclever** [comment]
To summarize

cefclient.exe without args prints correctly the pdf. On the other hand with minimalexample code there is no args to comment out
I will try seek into code what can be done further tomorrow

**Reply by amaitland** [reply]
Again you need to isolate which command line argument causes the problem with **cefclient**.

Start with each argument individually

```
 cefclient.exe --no-sandbox
```

```
cefclient.exe --multi-threaded-message-loop
```

```
cefclient.exe --disable-site-isolation-trials
```

At this point there is no further need to test with the CefSharp minimal example.

**Reply by emitclever** [reply]
I have removed  --no-sandbox and kept the other two, prints correctly. Same thing happens when I print to Microsoft PDF (pdf exported from browser is ok) since I have no printer available right now.

 ![image](https://user-images.githubusercontent.com/11360303/131762154-33b0498f-aeb2-4f71-aa61-3305ac895b04.png)



Put back --no-sandbox added as args prints incorrectly (export on Microsoft PDF)
![image](https://user-images.githubusercontent.com/11360303/131763000-d9a04c58-5e32-4148-ae1e-e9e5024466e1.png)

![image](https://user-images.githubusercontent.com/11360303/131762339-86a9c02a-6ad4-4f70-9292-f4911e98cd1f.png)

sample results dotted lines misplaced printed with args --no-sandox

![image](https://user-images.githubusercontent.com/11360303/131762460-43fe7e33-4a92-4def-ad90-dab196cdff0f.png)

sample results dotted lines OK printed with args removed --no-sandox

![image](https://user-images.githubusercontent.com/11360303/131763325-d3fad636-08b3-4b49-90ec-688578636ba1.png)

**Reply by emitclever** [reply]
Attached pdf samples 
[test-OK.pdf](https://github.com/cefsharp/CefSharp/files/7095208/test-OK.pdf)
[test-NotOK.pdf](https://github.com/cefsharp/CefSharp/files/7095224/test-NotOK.pdf)

**Comment by amaitland** [comment]
Unfortunately enabling the sandbox is not practical when running from .net, it requires statically linking of cef_sandbox.lib.

https://bitbucket.org/chromiumembedded/cef/wiki/SandboxSetup#markdown-header-windows

In theory you could create a native exe and bootstrap the .Net runtime manually, additional changes to CefSharp would also be required. I've never found time to test the theory.

You can test in Chrome with the same command line argument, if the problem reproduces then you can raise a bug at https://bugs.chromium.org/p/chromium/issues/list

If it works in Chrome then hopefully this will be fixed when https://bugs.chromium.org/p/chromium/issues/detail?id=702993 is resolved. 

Practically speaking you might need to look into non CEF based browser alternative.

**Reply by amaitland** [reply]
Likely related https://bitbucket.org/chromiumembedded/cef/issues/2757/radio-buttons-checkboxes-and-buttons-not

**Reply by emitclever** [reply]
thanks amaitland for support, this is a little bit over my head to resolve. I wish you all the best.
